tests/firecrackerdrivertest.c

```diff
@@ -0,0 +1,302 @@
+
+    virConnectPtr conn = virConnectOpen("fc:///system");
+
+    if (conn == NULL)
+        ret = -1;
+
+    if (virConnectClose(conn) < 0)
+        ret = -1;
+
+    return ret;
+}
+
+static void
+downloadKernelAndRootFS(void)
+{
+    const char *arch = virArchToString(virArchFromHost());
+    g_autofree const char *kernel_url = g_strdup_printf(
+        "https://s3.amazonaws.com/spec.ccfc.min/img/quickstart_guide/%s/kernels/vmlinux.bin",
+        arch);
+    g_autofree const char *rootfs_url = g_strdup_printf(
+        "https://s3.amazonaws.com/spec.ccfc.min/img/quickstart_guide/%s/rootfs/bionic.rootfs.ext4",
+        arch);
+
+    g_autofree const char *download_kernel_path = g_strdup_printf(
+        "%s/firecrackerdriverdata/hello-vmlinux-test.bin",
+        abs_srcdir);
+    g_autofree const char *download_rootfs_path = g_strdup_printf(
+        "%s/firecrackerdriverdata/hello-rootfs-test.ext4",
+        abs_srcdir);
+
+    virCommand *cmd = virCommandNew("curl");
+
+    virCommandAddArg(cmd, "-fsSL");
+    virCommandAddArg(cmd, "-o");
+    virCommandAddArg(cmd, download_kernel_path);
+    virCommandAddArg(cmd, kernel_url);
+
+    if (virCommandRun(cmd, NULL) < 0) {
+        VIR_WARN("Error downloading test kernel from link: %s",
+                 kernel_url);
+        return;
+    }
+
+    virCommandFree(cmd);
+
+    cmd = virCommandNew("curl");
+
+    virCommandAddArg(cmd, "-fsSL");
+    virCommandAddArg(cmd, "-o");
+    virCommandAddArg(cmd, download_rootfs_path);
+    virCommandAddArg(cmd, rootfs_url);
+
+    if (virCommandRun(cmd, NULL) < 0) {
+        VIR_WARN("Error downloading test rootfs from link: %s",
+                 rootfs_url);
+        return;
+    }
+
+    VIR_DEBUG("Finished downloading");
+}
+
+static virConnectPtr test_conn;
+
+static int
+testInitiateConnection(const void *opaque G_GNUC_UNUSED)
+{
+    test_conn = virConnectOpen("fc:///system");
+
+    if (test_conn == NULL)
+        return -1;
+
+    return 0;
+}
+
+static int
+testDefineDomain(const void *opaque G_GNUC_UNUSED)
+{
+    const char *path = g_strdup_printf("%s/firecrackerdriverdata/%s.xml",
+                                      abs_srcdir,
+                                      "test_domain");
+
+    char *xml = NULL;
+    if (virFileReadAll(path, 10 * 1024 * 1024, &xml) < 0) {
+        return -1;
+    }
+
+    VIR_DEBUG("xml read: %s", xml);
+
+    if (virDomainDefineXML(test_conn, xml) == NULL) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+testCreateDomain(const void *opaque G_GNUC_UNUSED)
+{
+    virDomainPtr dom = NULL;
+    int state = VIR_DOMAIN_NOSTATE;
+    int max_retries = 500;
+
+    if ((dom = virDomainLookupByName(test_conn, "firecracker_domain")) == NULL) {
+        return -1;
+    }
+
+    if (virDomainCreate(dom) < 0) {
+        return -1;
+    }
+
+    // Usually firecracker is set into running state even if the vm isn't
+    // yet ready to receive other commands like 'SendCtrlAltDel'
+    g_usleep(2500 * 1000);
```

Comment:
should this be moved inside the loop? and if we do move it in the loop, the 2.5 seconds * 500 retries is too much -> ~20 miins

Reply:
yes, but for a faulty implementation, this will loop for 20 mins before failinf

Reply:
i think that in a faulty implementation, if the function doesn't return, in 30 seconds the test is aborted. And if we are stuck in the while loop, there will be only 100 retries which will be easily under a few seconds tops.